### ISSUE TITLE
LibTorch (the C++ library) unnecessarily eats up memory

### ISSUE THREAD

**Issue by Chlorie** [pytorch/pytorch#15277]
## ❓ Questions and Help

I was trying to make predictions using a neural network and save the result tensors for later use. The tensors are just 65 dimensional vectors so it should not take much memory. But every time I do a forward pass and save the result, 3MB of memory is taken. I'm guessing that when I save the results, the parameters in the network are also kept in case I want to perform back propagation on that result tensor, but I'm not going to do that anyways. I tried setting the output tensor to not requiring gradient, and tried cloning the tensor into another one, but neither approach worked... Is it a bug in the code base or more possibly have I done anything wrong?

### Example code snippet for replicating the issue
    #include <vector>
    #include <torch/torch.h>

    struct Net :torch::nn::Module
    {
        torch::nn::Linear dense_1{ nullptr };
        torch::nn::Linear dense_2{ nullptr };
        Net()
        {
            dense_1 = register_module("dense_1", torch::nn::Linear(64, 10000));
            dense_2 = register_module("dense_2", torch::nn::Linear(10000, 65));
        }
        torch::Tensor forward(torch::Tensor input)
        {
            input = relu(dense_1->forward(input));
            input = dense_2->forward(input);
            return input;
        }
    };

    int main()
    {
        std::vector<torch::Tensor> outputs;
        Net net;
        net.eval();
        while (true) // Wait until memory usage blows up
        {
            const torch::Tensor input = torch::randn({ 1,64 }); // Generate some random data just for testing
            const torch::Tensor output = net.forward(input);
            outputs.push_back(output); // The output tensor isn't big, but since the weights of the dense layers are still kept, memory usage grows wildly
        }
    }

**Comment by Chlorie** [comment]
The network presented above is of course not the one I'm using, it's just for stating the problem.

**Comment by goldsborough** [comment]
Hi, thanks for the issue. This issue isn't really specific to the C++ frontend. If you run the equivalent of your code in Python:

```py
import torch

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.dense_1 = torch.nn.Linear(64, 10000)
        self.dense_2 = torch.nn.Linear(10000, 64)

    def forward(self, input):
        input = torch.relu(self.dense_1.forward(input))
        input = self.dense_2.forward(input)
        return input

outputs = []
net = Net()
net.eval()
while True:
    input = torch.randn(1, 64)
    output = net.forward(input)
    outputs.append(output)
```

and observe the memory usage of this program it will increase just the same.

The reason, as you state it, is indeed that you are saving the entire autograd history associated with `output` each time. You can avoid this by detaching the output from its autograd graph before saving it. Use `output.detach()` to do this in either C++ or Python. This should decrease the amount of memory used by each stored tensor.

**Comment by soumith** [comment]
closed via Peter's comment above.

**Comment by IgorAherne** [comment]
I have the following model:
`// Encoder network, which maps the input graph to a latent space encoding.
class VAE_EncoderImpl : public torch::nn::Module {
public:
    VAE_EncoderImpl(const std::vector<int>& sizes, int latent_size){
        for(int i=0; i<(int)sizes.size()-1; ++i){
            std::string name = std::string("encoderLinear")+std::to_string(i);
            auto lin = register_module(name, torch::nn::Linear(sizes[i], sizes[i + 1]));
            _layers.push_back(lin);
        }
        _linear_mu = register_module("encoderLinearMu", torch::nn::Linear(sizes.back(), latent_size));
        _linear_logvar = register_module("encoderLinearLogvar", torch::nn::Linear(sizes.back(), latent_size));
    }//end ctor


    std::tuple<torch::Tensor, torch::Tensor> forward(torch::Tensor x){
        // Use the encoder's layers to map the input graph to a hidden representation.
        for(int i=0; i<(int)_layers.size(); ++i){
            x = torch::silu(_layers[i]->forward(x));
        }
        // Use the encoder's mu and logvar layers to map the hidden representation to a latent space encoding.
        // Notice that we don't activate them.
        auto mu = _linear_mu->forward(x);
        auto logvar = _linear_logvar->forward(x);

        return std::make_tuple(mu, logvar);// Return the mu and logvar tensors.
    }

private:
    std::vector<torch::nn::Linear> _layers;
    torch::nn::Linear _linear1{nullptr};
    torch::nn::Linear _linear2{nullptr};
    torch::nn::Linear _linear3{nullptr};
    torch::nn::Linear _linear_mu{nullptr};
    torch::nn::Linear _linear_logvar{nullptr};
};
TORCH_MODULE(VAE_Encoder);`

Provide the code to save and load it, keeping in mind that layers are in a vector
